logstash-core/src/main/java/org/logstash/settings/DeprecatedAlias.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.logstash.settings;
+
+import co.elastic.logstash.api.DeprecationLogger;
+import org.apache.logging.log4j.LogManager;
+import org.logstash.log.DefaultDeprecationLogger;
+
+import java.util.function.Predicate;
+
+/**
+ * A <code>DeprecatedAlias</code> provides a deprecated alias for a setting, and is meant
+ * to be used exclusively through @see org.logstash.settings.SettingWithDeprecatedAlias#wrap()
+ * */
+final class DeprecatedAlias<T> extends SettingDelegator<T> {
+    private final DeprecationLogger deprecationLogger = new DefaultDeprecationLogger(LogManager.getLogger(DeprecatedAlias.class));
+
+    private SettingWithDeprecatedAlias<T> canonicalProxy;
+
+    protected DeprecatedAlias(String name, T defaultValue, boolean strict, Predicate<T> validator) {
+        super(name, defaultValue, strict, validator);
+    }
+
+    DeprecatedAlias(SettingWithDeprecatedAlias<T> canonicalProxy, String aliasName) {
+        super(canonicalProxy.getCanonicalSetting().clone().deprecate(aliasName));
+        this.canonicalProxy = canonicalProxy;
+    }
```

Comment:
Because `Setting#deprecate` returns a new `Setting` that is _separate_, we don't have access to its coersion or class-specific overrides.

To solve this, we can include the renaming in what this delegate is intercepting:

```suggestion
    final String aliasName;

    DeprecatedAlias(SettingWithDeprecatedAlias<T> canonicalProxy, String aliasName) {
        super(canonicalProxy.getCanonicalSetting().clone());
        this.canonicalProxy = canonicalProxy;
        this.aliasName = aliasName;
    }
    
    @Override
    String getName() {
        return this.aliasName;
    }

    @Override
    public void format(List<String> output) {
        // canonical setting is responsible for including info on deprecated alias
    }
```

Reply:
Nice point, but `Setting.deprecate` is invoked also by Ruby code (https://github.com/elastic/logstash/pull/15679/files#diff-291759c068472d4dd44be74d9d757c239e6e457b105df0f5d58218cc6a2b507fR849) so the deprecate instead of just create a new instance of `Setting` should clone itself, change the name of the clone to the alias and return the clone.

Reply:
Fixed with 09587199db80ea63a787df4ad1e36113208a1a24